text/0000-cargo-libstd-awareness.md

```diff
@@ -0,0 +1,154 @@
+
+ - Crate/symbol names are frobbed so that packages with the overlapping names don't conflict
+
+
+However rather than passing in libstd and its deps, Cargo lets the compiler look for them as need in
+the compiler's sysroot [specifically `<sysroot>/lib/<target>`]. This is quite coarse in comparison,
+and we loose all the advantages of the previous method:
+
+ - Packages may link or not link against libs in that directory as they please, with Cargo being
+   none the wiser.
+
+ - Cargo-built crates with the same name as those in there will collide, as the sysroot libs don't
+   have their names frobbed.
+
+ - Cross compiling may fail at build-time (as opposed to the much shorter
+   "gather-dependencies-time") because of missing packages
+
+
+Cargo doesn't look inside the sysroot to see what is or isn't there, but it would hardly help if it
+did, because it doesn't know what any package needs. Assuming all packages need libstd, for example,
+means Cargo just flat-out won't build freestanding packages that just use libcore on a platform that
+doesn't support libstd.
+
+For an anecdote: in https://github.com/RustOS-Fork-Holding-Ground I tried to rig up Cargo to cross
+compile libstd for me. Since I needed to use an unstable compiler anyways, it was possible in
+principle to build absolutely everything I needed with the same `rustc` version. Because of some
+trouble with Cargo and target JSONs, I didn't use a custom target specification, and just used
+`x86_64-gnu-linux`, meaning that depending on platform I was compiling on, I may or may have been
+cross-compiling. In the case where I wasn't, I couldn't complete the build because `rustc`
+complained about the libstd I was building overlapping with the libstd in the sysroot.
+
+For these reasons, most freestanding projects I know of avoid Cargo altogether, and just include
+submodule rust and run make in that. Cargo can still be used if one manages to get the requisite
+libraries in the sysroot. But this is a tedious operation that individual projects shouldn't need to
+reimplement, and one that has serious security implications if the normal libstd is modified.
+
+The fundamental plan proposed in this RFC is to make sure that anything Cargo builds never blindly
+links against libraries in the sysroot. This is achieved by making Cargo aware of all dependencies,
+including those libstd or its backing crates. That way, these problems are avoided.
+
+For the record, I first raised this issue [here](https://github.com/rust-lang/Cargo/issues/1096).
+
+
+# Detailed design
+
+The only new interface proposed is a boolean field in `Cargo.toml` specifying that the package does
+not depend on libstd by default. Note that this is technically orthogonal to Rust's `no_std`, as one
+might want to `use` their own build of libstd by default, or implicitly depend on it but not
+glob-import the prelude. To disambiguate, this field is called `implicit-deps`; please, go ahead and
```

Comment:
I see. Somebody beginning Rust, or beginning just even unstablized Rust, has no idea what dependencies would normally be implicit. One the other hand `implicit-std` sounds like one is just getting access to std, since Cargo normally doesn't allow one to `extern` transitive dependencies lest a direct dependency changes *its* dependencies. A name like `implicit-std-and-its-deps` would be the most clear, but my is it wordy.

Another option is to change it so that `implicit-std = true` only gives access to std itself, and dependencies on the crates behind the facade must always be explicit. This would also help on the off-chance that we want to version unstable std independently from its backing crates. Unfortunately, it also breaks backwards compatibility with existing packages, but only those using unstable Rust.